04-linear-chain.Rmd

```diff
@@ -0,0 +1,576 @@
+
+For instance, in this dummy `LinearChain` workflow below, after the task `stepA` is called, its task output `out` can be accessed at the workflow level as inputs for the task `stepB` in the form of `stepA.out`. This example is adapted from [OpenWDL Docs](https://docs.openwdl.org/en/latest/WDL/Linear_chaining/)' introduction to linear chain, which has more [introductory examples and variations of writing linear chains](https://docs.openwdl.org/en/latest/WDL/add_plumbing/).
+
+```         
+version 1.0
+workflow LinearChain {
+  input {
+    File firstInput
+  }
+  call stepA { 
+    input: 
+      in = firstInput
+  }
+  call stepB {
+   input:
+     in = stepA.out
+  }
+}
+
+task stepA {
+  input {
+    File in
+  }
+  command <<<
+    programA I=~{in} O=outputA.ext
+  >>>
+  output { 
+    File out = "outputA.ext"
+  }
+}
+
+task stepB {
+  input {
+    File in
+  }
+  command <<<
+    programB I=~{in} O=outputB.ext
+  >>>
+  output { 
+    File out = "outputB.ext"
+  }
+}
+```
+
+Now let's see how this applies to our workflow: we will take the output of `BwaMem` task into the input of the `MarkDuplicates` task. Recall that in the `BwaMem` task, the **output** section of the task is the following:
+
+```         
+output {
+    File analysisReadyBam = "~{base_file_name}.aligned.bam"
+    File analysisReadySorted = "~{base_file_name}.sorted_query_aligned.bam"
+  }
+```
+
+The File variables `analysisReadyBam` and `analysisReadySorted` can now be accessed anywhere in the workflow block after the `BwaMem` task as `BwaMem.``analysisReadyBam` and `BwaMem.analysisReadySorted`, respectively. Therefore, we call our next task `MarkDuplicates` with input `BwaMem.analysisReadySorted`:
+
+```         
+workflow minidata_mutation_calling_v1 {
+  input {
+    # Sample info
+    File sampleFastq
+    # Reference Genome information
+    File ref_fasta
+    File ref_fasta_index
+    File ref_dict
+    File ref_amb
+    File ref_ann
+    File ref_bwt
+    File ref_pac
+    File ref_sa
+    File dbSNP_vcf
+    File dbSNP_vcf_index
+    File known_indels_sites_VCFs
+    File known_indels_sites_indices
+    File af_only_gnomad
+    File af_only_gnomad_index
+    File annovarTAR
+    String annovar_protocols
+    String annovar_operation
+    String ref_name
+  }
+    
+  # Map reads to reference
+  call BwaMem {
+    input:
+      input_fastq = sampleFastq,
+      ref_fasta = ref_fasta,
+      ref_fasta_index = ref_fasta_index,
+      ref_dict = ref_dict,
+      ref_amb = ref_amb,
+      ref_ann = ref_ann,
+      ref_bwt = ref_bwt,
+      ref_pac = ref_pac,
+      ref_sa = ref_sa
+  }
+  
+  call MarkDuplicates {
+    input:
+      input_bam = BwaMem.analysisReadySorted
+  }
+}
+```
+
+## Writing `MarkDuplicates` task
+
+Of course, the task `MarkDuplicates` hasn't been written yet! Let's go through it together:
+
```

Comment:
Maybe we first write the second task and then show how to string it into a linear chain. So we could say something like " According to our workflow plan the next step would be to mark duplicate reads in our aligned BAM file generated by the BwaMem task. Before we can add this second task let us first write out our task..."

Reply:
I like introducing how to create linear chains before writing the task because the former is the focus of the chapter...